source/binding/Bitstream.cpp

```diff
@@ -679,9 +822,178 @@ ConstantValue Bitstream::evaluateTarget(const StreamingConcatenationExpression&
+            }
+
+            auto width = *right;
+            if (arrayType.hasFixedRange() &&
+                (bitwidth_t)width > arrayType.getFixedRange().width()) {
+                addDiag(diag::RangeWidthTooLarge, rightRange) << width << arrayType;
+                return false;
+            }
+
+            if (left && arrayType.hasFixedRange()) {
+                auto selection = RangeSelectExpression::getIndexedRange(
+                    kind == WithRangeKind::IndexedUp, *left, width,
+                    arrayType.getFixedRange().isLittleEndian());
+                auto bound = selection.right == *left ? selection.left : selection.right;
+                if (!arrayType.getFixedRange().containsPoint(bound)) {
+                    auto& diag = addDiag(diag::BadRangeExpression, errorRange);
+                    diag << selection.left << selection.right;
+                    diag << arrayType;
+                    return false;
+                }
+            }
+        }
+        else {
+            // Reference RangeSelectExpression::fromSyntax for similar constraints of simple
+            // part-select
+            if (!validateIndex(*right, rightRange))
+                return false;
+
+            if (left) {
+                ConstantRange selection = { *left, *right };
+                if (selection.width() > 1 &&
+                    selection.isLittleEndian() != (arrayType.hasFixedRange()
+                                                       ? arrayType.getFixedRange().isLittleEndian()
+                                                       : false)) {
+                    auto& diag = addDiag(diag::SelectEndianMismatch, errorRange);
+                    diag << arrayType;
+                    return false;
+                }
+            }
+        }
+    }
+
+    return true;
+}
+
+optional<int32_t> Bitstream::withRangeWidth(WithRangeKind kind, optional<int32_t> left,
+                                            optional<int32_t> right) {
+    if (kind == WithRangeKind::Bit)
+        return 1;
+    else if (kind == WithRangeKind::IndexedUp || kind == WithRangeKind::IndexedDown)
+        return right;
+    else if (left && right) {
+        ConstantRange range = { *left, *right };
+        return range.width();
+    }
+    return std::nullopt;
+}
+
+optional<ConstantRange> Bitstream::evaluateWith(
+    const Type& arrayType, const StreamingConcatenationExpression::WithExpression& with,
+    EvalContext& context) {
+
+    auto evalInt = [&](const Expression& expr) -> optional<int32_t> {
+        auto cs = expr.eval(context);
+        if (!cs)
+            return std::nullopt;
+        auto index = cs.integer().as<int32_t>();
+        if (!index) {
+            context.addDiag(diag::ConstEvalArrayIndexInvalid, expr.sourceRange) << cs << arrayType;
+        }
+        return index;
+    };
+
+    auto cl = evalInt(*with.left);
+    if (!cl)
+        return std::nullopt;
+    decltype(cl) cr = std::nullopt;
```

Comment:
Using decltype() here seems not necessary, the full type is not long or hard to spell out.

Reply:
Yes, I replace it with optional<int32_t>.